ext/OptimizationDIExt.jl

```diff
@@ -0,0 +1,181 @@
+module OptimizationDIExt
+       
+
+import OptimizationBase, OptimizationBase.ArrayInterface
+import OptimizationBase.SciMLBase: OptimizationFunction
+import OptimizationBase.LinearAlgebra: I
+import DifferentiationInterface
+import DifferentiationInterface: prepare_gradient, prepare_hessian, prepare_jacobian, gradient!!, hessian!!, jacobian!!
+using ADTypes
+
+function OptimizationBase.instantiate_function(f, x, adtype::ADTypes.AbstractADType, p = SciMLBase.NullParameters(), num_cons = 0)
+    _f = (θ, args...) -> first(f.f(θ, p, args...))
+
+    if f.grad === nothing
+        extras_grad = prepare_gradient(_f, adtype, x)
+        function grad(res, θ)
+            gradient!!(_f, res, adtype, θ, extras_grad)
```

Comment:
there is no guarantee that `gradient!!` actually overwrites `res`, so you need to do

```julia
res .= gradient!!(_f, res, adtype, θ, extras_grad)
```

See https://gdalle.github.io/DifferentiationInterface.jl/DifferentiationInterface/stable/overview/#Variants for details on the bang-bang syntax

Reply:
the same goes for all the other operators

Reply:
Shouldn't it rather be
```julia
res = gradient!!(_f, res, adtype, θ, extras_grad)
```
If `res` is immutable, you can't do `res .= ...`; and if it is mutable, hopefully `gradient!!` already updated it.

Reply:
You're right that this is the more generic and efficient version. I didn't check to see if Optimization _requires_ an in-place update of `res` or not. If it does, you need `res .= ...`. If you can get away with re-assigning `res`, all the better (but beware of type instability, maybe name it differently)

Reply:
The reason I mention type-instability is that when you call `res_after = gradient!!(f, res_before, ...)`, the returned `res_after` might be something very different from `res_before`, typically a lazy tangent like a `FillArray`.

Reply:
This assumes mutability and in-place. So the original suggestions make sense 👍

Reply:
If `gradient!!` requires `res .= ...`, IMO it does not satisfy the bang bang convention and hence should be considered broken. The bang bang convention is supposed to allow you to write generic code that works for both mutable and immutable containers, and that you can only do with the `res = f!!(res)` pattern.

Reply:
That being fair I don't see where I deviate from the bang bang convention. There's no clear way to decide whether an object is mutable, but not mutating it and returning a new one is always a valid choice

Reply:
I edited the warning to avoid suggesting that there is a connection
https://github.com/gdalle/DifferentiationInterface.jl/pull/167

Reply:
But note that BangBang.jl will not always mutate the input even when it can:
```julia
julia> using BangBang, StaticArrays

julia> src = [2.0, 3.0];

julia> dst_before = zero(src);

julia> dst_after = broadcast!!(abs2, dst_before, src)
2-element Vector{Float64}:
 4.0
 9.0

julia> dst_before  # it was mutated
2-element Vector{Float64}:
 4.0
 9.0

julia> src_static = MVector(2.0, 3.0);

julia> dst_before_static = zero(src_static);

julia> dst_after_static = broadcast!!(abs2, dst_before_static, src_static)
2-element MVector{2, Float64} with indices SOneTo(2):
 4.0
 9.0

julia> dst_before_static. # it wasn't mutated but could have been
2-element MVector{2, Float64} with indices SOneTo(2):
 0.0
 0.0
```

Reply:
IMO the point of the bang bang convention is being able to write generic code that operates inplace whenever possible and out-of-place otherwise: For mutable containers, I can use `f!(x)` and for immutable containers `x = f(x)`; but without knowing whether `x` is mutable or not the bang bang pattern allows me to use the generic `x = f!!(x)` which does `x = (f!(x); x)` for mutable `x` and  `x = f(x)` for immutable `x`. There's no reason for writing `x .= f!!(x)` with the bang bang convention since this will only work for mutable arrays and in that case it would be easier to just call `f!(x)`.

Reply:
The example above just shows you that even BangBang itself does not guarantee they will mutate mutable arrays

Reply:
So if they don't, we can't either

Reply:
My point stands: if the inventors of this BangBang convention cannot guarantee mutation on one of the most common mutable array types, I won't claim to. Even if we fix this bug, there may be others waiting.
So working from this principle, we cannot be one hundred percent sure that `f!(x)` will actually mutate `x`. All we can say is that it might. My take on this is that doing `x .= f!!(x)` ensures an error will occur with a non-mutable array, instead of a silent "do-nothing". It is slower, that is true, but I would assume for most practical purposes this is dwarved by the actual gradient computation

Reply:
> That's a core Julia convention, so I think any violation of this principle is 100% a bug.

This is an ambiguous point for me, and for others too according to this [Discourse thread](https://discourse.julialang.org/t/why-is-bang-indicating-argument-mutation-a-convention-but-not-enforced/49783/3?u=gdalle). The [documentation on functions](https://docs.julialang.org/en/v1/manual/functions/#man-argument-passing) states:

> _As a common convention in Julia (not a syntactic requirement), such a function would [typically be named f!(x, y)](https://docs.julialang.org/en/v1/base/punctuation/#man-punctuation) rather than f(x, y), as a visual reminder at the call site that at least one of the arguments (often the first one) is being mutated._

Okay then. Is the first argument "being mutated" in the following function?

```julia
function double!(storage; mutate)
    if mutate
        storage .*= 2
        return storage
    else
        return 2 .* storage
    end
end
```
If you agree that it is, then how about in this one?

```julia
function gradient!(grad, x; backend)
    if backend == :ForwardDiff
        ForwardDiff.gradient!(grad, f, x)
        return grad
    elseif backend == :Zygote
        return Zygote.gradient(f, x)
    end
end
```

This second example is exactly what happens in DifferentiationInterface. So I would say that functions ending in `!` are _not forced to guarantee mutation for every possible path taken by the control flow_. And even if they were, there is absolutely no way to enforce that, as demonstrated by the BangBang example (which you deemed a bug).

> And broadcasting puts quite a bit of stress on the compiler, so personally I'd at least use copyto!(x, ff!(x)).

Point taken, I didn't know there was a difference.

> But also from your last comment it's still not clear to me why you would ever use that pattern. If you assume/require mutability, just call f!(x). It's faster, more direct, and seems what you actually intended to use.

Because the primary focus of `gradient!!(grad, backend, f, x)` is not to _update `grad`_, it is to _provide storage to the backend in case it is needed for performance_.
My perspective from the start was that some backends can use that storage, and others will discard it, so we might as well allow the user to provide it in both cases.

Additionally, the latest AD backend [Tapir.jl](https://github.com/withbayes/Tapir.jl) was designed using this exact same philosophy of "possibly mutate, always return". Perhaps @willtebbutt can back me up here.

Reply:
Oh, and I forgot the main motivation behind this initial choice: it ensures a common interface for `pushforward!!` and `pullback!!` regardless of the types of the inputs and outputs. 

If we want to go back to a "simple-bang" convention with enforced mutation, we need to sprinkle `copyto!` everywhere. This will work for `AbstractArray`s, but for custom structures it will fail, thus closing the door on people who want to use our code for more generic purposes.

Reply:
This discussion and talking with @adrhill actually made me reconsider. It's gonna suck testing-wise but I'll open a PR to try and enforce mutation. We'll see how it goes

Reply:
> Additionally, the latest AD backend [Tapir.jl](https://github.com/withbayes/Tapir.jl) was designed using this exact same philosophy of "possibly mutate, always return". Perhaps @willtebbutt can back me up here, because that's one of the places I drew inspiration from.

I'll comment on `Tapir.jl`'s usage of the `!!` convention: `Tapir.rrule!!(f_, x_...)` will mutate iff `f(x...)` mutates. The same applies to `Tapir.value_and_pullback!!` and `Tapir.value_and_gradient!!`. Largely speaking, if write
```julia
y = f(x...)
```
the appropriate `rrule!!`-version of the statement would be
```julia
y_ = rrule!!(f_, x_...)
```
(`y` and `y_` are slightly different data structures to each other, but morally correspond to one another -- the same applies to `f` / `f_` and `x` / `x_`).

Hope this is useful!

Reply:
Okay, I had not understood that! I thought it also had to do with the mutability of the storage provided for the VJP. But I guess it is connected since they have to have the same type as `x`, and `f` can only mutate a mutable `x`

Reply:
In any case, I have started a PR to go back to the good old "mutate always" paradigm
https://github.com/gdalle/DifferentiationInterface.jl/pull/169

Reply:
Hmm you make a good point.

For `rrule!!`, which is a regular state-free function, what I say above is correct. For rules produced by `build_rrule!!`, however, you're correct that there is internal state that gets modified whenever you run the rule or the pullback that it produces.